Smarter state fields

dpsutton opened the pull request:
Fixes https://github.com/metabase/metabase/issues/2735

Removes marking fields as `:type/State` based on their name and introduces a new fingerprinter to check text values for full state names or abbreviations. A fingerprinting classifier then uses this, and if a column is marked at over 70% state-like it will be marked as a state.

Some intricacy as there are four classifiers and they are very order dependent. I've added the ability for fingerprint classifiers to override previous classifiers if the original field value didn't have a value. In the future i think this should get even smarter and we can compute possible classifications and then resolve the best one, or perhaps let downstream classifiers subclassify (ie `:text/Category` move to something that derives from category).

Lots of expectations -> clojure.test. 

Points to consider
- is the `metabase.util/state?` too simple to be in util and perhaps should be in the fingerprinter namespace. Is it too simple? should it have canadian states? allow for locales to load an edn file from resources with state names, etc.
- i debated moving the fingerprinting classifiers earlier but this seems pretty fraught with unintended consequences
```clojure
(def ^:private classifiers
  "Various classifier functions available. These should all take two args, a `FieldInstance` and a possibly `nil`
  `Fingerprint`, and return `FieldInstance` with any inferred property changes, or `nil` if none could be inferred.
  Order is important!"
  [name/infer-and-assoc-special-type
   category/infer-is-category-or-list
   no-preview-display/infer-no-preview-display
   text-fingerprint/infer-special-type])
```
- debated adding metadata to these classifiers of what they change, what they are allowed to change and have the reduction use that information but it got a bit hairy. I think something like this should be done and more smarts can go into classification but was scared of too many surfaces for regressions at one time.
- i debated adding the original field into the signature of these but kinda backed and snuck it in through metadata. not sure if there are strong feelings about that one way or another
- considered adding a check that if the original field had no `:special_type` but a classifier had added one, then only add one in the fingerprinters if the new special type was a derivation of the one already selected. I abandoned this as `:type/State` does derive from category but json, url, and email do not for obvious reasons.
- one change from this is that if anything was previously marked as `:type/Category` (and this happens for almost every field with fewer than 100 distinct items) it would never be marked as url, json, or email even if 100% of the values matched these fingerprints. Now we take the word of the fingerprinters.
- introduced a new threshold for states. kept the 95% threshold for email, url, and json but figured with states this is a bit stringent. Had in mind US territories, washington DC, Canadian provinces, etc. so wanted a bit more wiggle room for matching a state.

dpsutton committed:
Mark state fields by fingerprint fixes #2735 [ci drivers]

dpsutton committed:
Expectations -> clojure.test

dpsutton committed:
fix tests for pg, ns sort, and docstring lint

dpsutton committed:
minor alignment in test

dpsutton committed:
Allow fingerprint classifiers to override prev classifier special_type

if another classifier threw on a special type due to name or the
number of items in the collection, this should override it as it is
introspecting values.

States previously were marked ":type/State" based on the column
name. Now use a fingerprinter looking for state names. Owing to the
fact that there might be more things in the state column (US
territory, canadian states, Washington DC, etc) lowered the threshold
for recognizing a state column from 95% of values down to 70% of
values.

mazameli commented on src/metabase/util.clj:
```diff
@@ -94,6 +94,21 @@
+(defn state?
+  "Is `s` a state string?"
+  ^Boolean [^String s]
+  (contains? #{"alabama" "alaska" "arizona" "arkansas" "california" "colorado" "connecticut" "delaware"
```
Is this too USA-centric? E.g., what if I'm a Brazilian user and my `State` field contains "acre" "bahia" "goias" "roraima" "sao paulo" etc.? We have users in many countries, many of which are made up of states (Brazil, Australia, Germany, India …).

dpsutton replied:
do we have any UI bits that can show maps of other countries? I raised this issue in the description with a possibility of looking for some config file for states we could check. But do we have existing or planned UI bits to take advantage of it? I'm also not sure where `:type/State` would be used besides the USA map but i certainly want to make sure I haven't violated any assumptions.

paulrosenzweig replied:
You can upload your own custom maps, but AFAIK we don't have a way to label a map as the official "map of states" for your instance. Without a feature like that marking non-US states as `type/State` might lead to a worse experience if it often displays as a blank map of the US. I haven't tested it, but that would be my guess.

That said, it'd be sweet to bundle more maps with Metabase. Maybe when we do that it'd make sense to have separate types for each country's states/regions? That does inflate the type system a lot though.

dpsutton replied:
ok. it sounds like non-us-state recognition is a future endeavor then? perhaps in the future we can have some kind of `:type.State/US`, `:type.State/Brazil` or some similar notion but that day is not today.

mazameli replied:
Greeeaaat point, Paul. I lost sight of what marked a field as `type/State` actually does currently. 🤦‍♂️

sbelak commented on src/metabase/sync/analyze/classifiers/text_fingerprint.clj:
```diff
@@ -8,45 +8,58 @@
-        (seq percent-key->special-type)))
+        percent-key->special-type))
+
+(defn- can-edit-special-type?
```
Can you document/explain the algorithm (tagging via metadata) here

dpsutton replied:
good catch! will do

sbelak reviewed (approved):
I agree we need a more comprehensive solution for figuring out what values we can overwrite. There's #12574 tracking that.

Similarly at some point we should rewrite the entire classification logic as some form of weighted voting (ensamble/blackboard) rather than a linear chain of classifiers that potentially overwrite one another. Overwriting the generic `:type/Category` is correct. _Use the more specific type_ should be a decent heuristic, but is currently hampered by `:type/Category` being derived from `:type/Special` (which as far as I can tell could be safely removed altogether). 

I'm not a huge of meta for this as it's relatively fragile, but it'a not a deal breaker. 

This might be beyond the scope of this PR, but we should revisit the decision to count nils towards the classification threshold. I think skipping nils in https://github.com/metabase/metabase/blob/master/src/metabase/sync/analyze/fingerprint/fingerprinters.clj#L243 would make more sense. 

How does the migration look for this, especially for folks who currently have miss-classified `:state` fields?

sbelak commented on src/metabase/util.clj:
```diff
@@ -94,6 +94,21 @@
+               "al" "mt" "ak" "ne" "az" "nv" "ar" "nh" "ca" "nj" "co" "nm" "ct" "ny" "de" "nc" "fl" "nd"
+               "ga" "oh" "hi" "ok" "id" "or" "il" "pa" "in" "ri" "ia" "sc" "ks" "sd" "ky" "tn" "la" "tx"
+               "me" "ut" "md" "vt" "ma" "va" "mi" "wa" "mn" "wv" "ms" "wi" "mo" "wy"}
+             (when s (str/lower-case s))))
```
I'd lift the `when` to include `contains?` and/or use `some->>`.

Also not sure which is the better tradeoff: keeping the set in memory permanently, or paying the cost to (potentially) recreate it on every invocation. I'm leaning towards the former, but this should be tested.

dpsutton replied:
the clojure compiler will emit this into the constant pool i believe. its only made into a set on reading

dpsutton replied:
```clojure
(decompile (defn state?
  "Is `s` a state string?"
  ^Boolean [^String s]
  (contains? #{"alabama" "alaska" "arizona" "arkansas" "california" "colorado" "connecticut" "delaware"
               "florida" "georgia" "hawaii" "idaho" "illinois" "indiana" "iowa" "kansas" "kentucky" "louisiana"
               "maine" "maryland" "massachusetts" "michigan" "minnesota" "mississippi" "missouri" "montana"
               "nebraska" "nevada" "new hampshire" "new jersey" "new mexico" "new york" "north carolina"
               "north dakota" "ohio" "oklahoma" "oregon" "pennsylvania" "rhode island" "south carolina"
               "south dakota" "tennessee" "texas" "utah" "vermont" "virginia" "washington" "west virginia"
               "wisconsin" "wyoming"
               "al" "mt" "ak" "ne" "az" "nv" "ar" "nh" "ca" "nj" "co" "nm" "ct" "ny" "de" "nc" "fl" "nd"
               "ga" "oh" "hi" "ok" "id" "or" "il" "pa" "in" "ri" "ia" "sc" "ks" "sd" "ky" "tn" "la" "tx"
               "me" "ut" "md" "vt" "ma" "va" "mi" "wa" "mn" "wv" "ms" "wi" "mo" "wy"}
             (when s (str/lower-case s)))))

// Decompiling class: user$state_QMARK_
import clojure.lang.*;

public final class user$state_QMARK_ extends AFunction
{
    public static final Var const__0;
    public static final AFn const__1;
    public static final Var const__2;

    public static Object invokeStatic(final Object s) {
        final IFn fn = (IFn)user$state_QMARK_.const__0.getRawRoot();
        final AFn const__1 = user$state_QMARK_.const__1;
        if (s != null) {
            if (s != Boolean.FALSE) {
                final Object invoke = ((IFn)user$state_QMARK_.const__2.getRawRoot()).invoke(s);
                return fn.invoke(const__1, invoke);
            }
        }
        final Object invoke = null;
        return fn.invoke(const__1, invoke);
    }

    @Override
    public Object invoke(final Object s) {
        return invokeStatic(s);
    }

    static {
        const__0 = RT.var("clojure.core", "contains?");
        const__1 = PersistentHashSet.create("rhode island", "delaware", "ct", "ohio", "new york", "colorado", "nh", "wi", "la", "california", "md", "wisconsin", "massachusetts", "ne", "alaska", "al", "pennsylvania", "oklahoma", "texas", "mo", "vt", "north carolina", "idaho", "nevada", "new jersey", "indiana", "west virginia", "florida", "new mexico", "nc", "ky", "maine", "sc", "kentucky", "id", "arkansas", "utah", "co", "connecticut", "vermont", "ca", "mi", "ut", "wv", "kansas", "ma", "ia", "wa", "ms", "south dakota", "fl", "il", "oh", "south carolina", "tn", "or", "de", "ri", "hi", "wy", "iowa", "maryland", "ga", "hawaii", "montana", "wyoming", "mt", "tx", "missouri", "oregon", "georgia", "nm", "mn", "north dakota", "virginia", "illinois", "ar", "ny", "sd", "ak", "ok", "alabama", "nv", "louisiana", "nebraska", "tennessee", "new hampshire", "washington", "pa", "va", "minnesota", "ks", "me", "nj", "arizona", "michigan", "in", "az", "mississippi", "nd");
        const__2 = RT.var("clojure.string", "lower-case");
    }
}

nil
```

yeah. and checking for nil versus checking the set for nil is probably not a huge difference.

sbelak replied:
Good catch. TIL -- I thought it only does that for small objects.

dpsutton commented on src/metabase/sync/analyze/classifiers/text_fingerprint.clj:
```diff
@@ -8,45 +8,58 @@
-  {:percent-json  :type/SerializedJSON
-   :percent-url   :type/URL
-   :percent-email :type/Email})
+  {:percent-json  [:type/SerializedJSON percent-valid-threshold]
```
i kinda agree. but got told otherwise for the SSH heartbeat. However, i think some tests use this so maybe it has a purpose

dpsutton committed:
Cleanup [ci drivers]

- remark about original field in metadata
- docstring on `can-edit-special-field?`
- cleanup and check for string in `state?`
- add tests that nils and other types don't break `state?`

dpsutton committed:
Include test file

dpsutton commented:
> How does the migration look for this, especially for folks who currently have miss-classified :state fields?

I don't have a good approach to this. I've been thinking about it but there's seemingly not enough information. My current thinking is that it is better to preserve a bad analysis than risk throwing away someone's manual customization. As it stands, there's no way to tell if someone marked something as State manually or if we inferred it. There is a ticket to add more information about this so we can include manual intervention into our classification.
